pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go

```diff
@@ -0,0 +1,335 @@
+			if err := r.client.Update(context.TODO(), clusterset); err != nil {
+				klog.Warningf("will reconcile since failed to remove Finalizer from ManagedClusterset %v, %v", clusterset.Name, err)
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if !utils.ContainsString(clusterset.GetFinalizers(), clustersetRoleFinalizerName) {
+		klog.Infof("adding Clusterrole Finalizer to ManagedClusterset %v", clusterset.Name)
+		clusterset.ObjectMeta.Finalizers = append(clusterset.ObjectMeta.Finalizers, clustersetRoleFinalizerName)
+		if err := r.client.Update(context.TODO(), clusterset); err != nil {
+			klog.Warningf("will reconcile since failed to add finalizer to ManagedClusterset %v, %v", clusterset.Name, err)
+			return ctrl.Result{}, err
+		}
+	}
+
+	//add clusterset clusterrole
+	adminRules := buildAdminRoleRules(clusterset.Name)
+	err = utils.ApplyClusterRole(r.kubeClient, utils.GenerateClustersetClusterroleName(clusterset.Name, "admin"), adminRules)
+	if err != nil {
+		klog.Warningf("will reconcile since failed to create/update clusterrole %v, %v", clusterset.Name, err)
+		return ctrl.Result{}, err
+	}
+	viewRules := buildViewRoleRules(clusterset.Name)
+	err = utils.ApplyClusterRole(r.kubeClient, utils.GenerateClustersetClusterroleName(clusterset.Name, "view"), viewRules)
+	if err != nil {
+		klog.Warningf("will reconcile since failed to create/update clusterrole %v, %v", clusterset.Name, err)
+		return ctrl.Result{}, err
+	}
+
+	r.syncClustersetMapper(clusterset.Name)
+
+	return ctrl.Result{}, nil
+}
+
+//cleanClusterSetResource clean the clusterrole
+//and delete clusterset related resource in clusterSetClusterMapper and clusterSetNamespaceMapper
+func (r *Reconciler) cleanClusterSetResource(clustersetName string) error {
+	//Delete clusterset clusterrole
+	err := utils.DeleteClusterRole(r.kubeClient, utils.GenerateClustersetClusterroleName(clustersetName, "admin"))
+	if err != nil {
+		klog.Warningf("will reconcile since failed to delete clusterrole. clusterset: %v, err: %v", clustersetName, err)
+		return err
+	}
+	err = utils.DeleteClusterRole(r.kubeClient, utils.GenerateClustersetClusterroleName(clustersetName, "view"))
+	if err != nil {
+		klog.Warningf("will reconcile since failed to delete clusterrole. clusterset: %v, err: %v", clustersetName, err)
+		return err
+	}
+
+	//Delete clusterset which in clustersetMapper
+	r.clusterSetClusterMapper.DeleteClusterSet(clustersetName)
+	r.clusterSetNamespaceMapper.DeleteClusterSet(clustersetName)
+
+	return nil
+}
+
+func (r *Reconciler) syncClustersetMapper(clustersetName string) error {
+	//List Clusterset related resource by utils.ClusterSetLabel
+	labelSelector := metav1.LabelSelector{MatchLabels: map[string]string{
+		clustersetutils.ClusterSetLabel: clustersetName,
+	}}
+	selector, err := metav1.LabelSelectorAsSelector(&labelSelector)
+	if err != nil {
+		return err
+	}
+	requiredClusterSetClusterMapper, err := r.generateClustersetClusterMapper(selector)
```

Comment:
i am not sure  why  you need generate this cache here. It is is find clusters in a clusterset. You can directly filter it, right?